src/query/block/column.go

```diff
@@ -133,154 +108,162 @@ func (c *colBlockIter) Next() bool {
-	col := c.columns[c.idx]
-	return ColStep{
-		time:   c.timeForStep,
-		values: col.Values,
-	}
-}
-
-func (c *colBlockIter) Close() { /*no-op*/ }
-
-// ColStep is a single column containing data from multiple series at a given time step
+// ColStep is a single column containing data from multiple
+// series at a given time step.
 type ColStep struct {
 	time   time.Time
 	values []float64
 }
 
-// Time for the step
+// Time for the step.
 func (c ColStep) Time() time.Time {
 	return c.time
 }
 
-// Values for the column
+// Values for the column.
 func (c ColStep) Values() []float64 {
 	return c.values
 }
 
-// NewColStep creates a new column step
+// NewColStep creates a new column step.
 func NewColStep(t time.Time, values []float64) Step {
 	return ColStep{time: t, values: values}
 }
 
-// NewColumnBlockBuilder creates a new column block builder
-func NewColumnBlockBuilder(
-	queryCtx *models.QueryContext,
-	meta Metadata,
-	seriesMeta []SeriesMeta) Builder {
-	return ColumnBlockBuilder{
-		enforcer: queryCtx.Enforcer.Child(cost.BlockLevel),
-		blockDatapoints: queryCtx.Scope.Tagged(
-			map[string]string{"type": "generated"}).Counter("datapoints"),
-		block: &columnBlock{
-			meta:       meta,
-			seriesMeta: seriesMeta,
-			blockType:  BlockDecompressed,
-		},
+func (c *colBlockIter) Current() Step {
+	col := c.columns[c.idx]
+	return ColStep{
+		time:   c.timeForStep,
+		values: col,
 	}
 }
 
-// AppendValue adds a value to a column at index
-func (cb ColumnBlockBuilder) AppendValue(idx int, value float64) error {
-	columns := cb.block.columns
-	if len(columns) <= idx {
-		return fmt.Errorf("idx out of range for append: %d", idx)
-	}
+// Close frees up resources held by the iterator.
+func (c *colBlockIter) Close() { /*no-op*/ }
 
-	r := cb.enforcer.Add(1)
-	if r.Error != nil {
-		return r.Error
-	}
+type colSeriesIter struct {
+	startIdx   int
+	endIdx     int
+	idx        int
+	meta       Metadata
+	seriesMeta []SeriesMeta
+	columns    []column
+	dps        ts.Datapoints
+}
 
-	cb.blockDatapoints.Inc(1)
+// SeriesMeta returns the metadata for each series in the block.
+func (it *colSeriesIter) SeriesMeta() []SeriesMeta { return it.seriesMeta }
 
-	columns[idx].Values = append(columns[idx].Values, value)
-	return nil
-}
+// SeriesCount returns the number of series.
+func (it *colSeriesIter) SeriesCount() int { return len(it.seriesMeta) }
 
-// AppendValues adds a slice of values to a column at index
-func (cb ColumnBlockBuilder) AppendValues(idx int, values []float64) error {
-	columns := cb.block.columns
-	if len(columns) <= idx {
-		return fmt.Errorf("idx out of range for append: %d", idx)
-	}
+// Current returns the current series for the block.
+func (it *colSeriesIter) Current() UnconsolidatedSeries {
+	var (
+		bounds = it.meta.Bounds
+		steps  = bounds.Steps()
+		start  = bounds.Start
+		step   = bounds.StepSize
+	)
 
-	r := cb.enforcer.Add(xcost.Cost(len(values)))
-	if r.Error != nil {
-		return r.Error
+	if it.dps == nil {
```

Comment:
Hm, is it also worth checking if the steps are enough?  I could see `it.dps[i]` causing out of range if `it.dps` is not long enough for `bounds.Steps()`?

Could just change this check to be `if it.dps == nil || len(it.dps) < steps { /*alloc*/ }` just to be safe perhaps?